File: src/main/哈希表/t290/WordPattern.java
package 哈希表.t290;

import java.util.HashMap;

/**
 * @auther: sotowang
 * @date: 2019/11/24 20:15
 */
public class WordPattern {
    public boolean wordPattern(String pattern, String str) {
        String[] strings = str.split(" ");
        if (pattern.length() != strings.length) {
            return false;
        }
        HashMap<Character, String> hashMap = new HashMap<>();
        for (int i = 0; i < pattern.length(); i++) {
            if (!hashMap.containsKey(pattern.charAt(i))) {
                if (hashMap.containsValue(strings[i])) {
                    return false;
                }
                hashMap.put(pattern.charAt(i), strings[i]);
            } else {
                if (!hashMap.get(pattern.charAt(i)).equals(strings[i])) {
                    return false;
                }
            }
        }
        return true;
    }

    public static void main(String[] args) {
        String pattern1 = "abba";
        String str1 = "dog cat cat dog";
        System.out.println(new WordPattern().wordPattern(pattern1, str1));
        String pattern2 = "abba";
        String str2 = "dog cat cat fish";
        System.out.println(new WordPattern().wordPattern(pattern2, str2));
    }
}

File: src/main/哈希表/t290/wordPattern.go
package main

import (
	"fmt"
	"strings"
)

func wordPattern(pattern string, s string) bool {
	ss := strings.Fields(s)
	pLen, sLen := len(pattern), len(ss)
	if pLen != sLen {
		return false
	}
	keyMap := make(map[byte]string)
	valueMap := make(map[string]byte)
	for i := 0; i < pLen; i++ {
		if _, ok := keyMap[pattern[i]]; !ok {
			if _, ok := valueMap[ss[i]]; ok {
				return false
			}
			keyMap[pattern[i]] = ss[i]
			valueMap[ss[i]] = pattern[i]
		} else if val, _ := keyMap[pattern[i]]; val != ss[i] {
			return false
		}
	}
	return true
}

func main() {
	s := "dog cat cat dog"
	pattern := "abba"
	fmt.Print(wordPattern(pattern, s))
}
